File: python/dataproc_templates/gcs/gcs_to_mongo.py
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Sequence, Optional, Any
from logging import Logger
import argparse
import pprint
from pyspark.sql import SparkSession

from dataproc_templates import BaseTemplate
import dataproc_templates.util.template_constants as constants
from dataproc_templates.util.argument_parsing import add_spark_options
from dataproc_templates.util.dataframe_reader_wrappers import ingest_dataframe_from_cloud_storage


__all__ = ['GCSToMONGOTemplate']


class GCSToMONGOTemplate(BaseTemplate):
    """
    Dataproc template implementing loads from GCS into MongoDB Database
    """

    @staticmethod
    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
        parser: argparse.ArgumentParser = argparse.ArgumentParser()

        parser.add_argument(
            f'--{constants.GCS_MONGO_INPUT_LOCATION}',
            dest=constants.GCS_MONGO_INPUT_LOCATION,
            required=True,
            help='Cloud Storage location of the input files'
        )
        parser.add_argument(
            f'--{constants.GCS_MONGO_INPUT_FORMAT}',
            dest=constants.GCS_MONGO_INPUT_FORMAT,
            required=True,
            help='Input file format (one of: avro,parquet,csv,json,delta)',
            choices=[
                constants.FORMAT_AVRO,
                constants.FORMAT_PRQT,
                constants.FORMAT_CSV,
                constants.FORMAT_JSON,
                constants.FORMAT_DELTA
            ]
        )
        add_spark_options(parser, constants.get_csv_input_spark_options("gcs.mongo.input."))
        parser.add_argument(
            f'--{constants.GCS_MONGO_OUTPUT_URI}',
            dest=constants.GCS_MONGO_OUTPUT_URI,
            required=True,
            help='GCS MONGO Output Connection Uri'
        )
        parser.add_argument(
            f'--{constants.GCS_MONGO_OUTPUT_DATABASE}',
            dest=constants.GCS_MONGO_OUTPUT_DATABASE,
            required=True,
            help='GCS MONGO Output Database Name'
        )
        parser.add_argument(
            f'--{constants.GCS_MONGO_OUTPUT_COLLECTION}',
            dest=constants.GCS_MONGO_OUTPUT_COLLECTION,
            required=True,
            help='GCS MONGO Output Collection Name'
        )
        parser.add_argument(
            f'--{constants.GCS_MONGO_OUTPUT_MODE}',
            dest=constants.GCS_MONGO_OUTPUT_MODE,
            required=False,
            default=constants.OUTPUT_MODE_APPEND,
            help=(
                'Output write mode '
                '(one of: append,overwrite,ignore,errorifexists) '
                '(Defaults to append)'
            ),
            choices=[
                constants.OUTPUT_MODE_OVERWRITE,
                constants.OUTPUT_MODE_APPEND,
                constants.OUTPUT_MODE_IGNORE,
                constants.OUTPUT_MODE_ERRORIFEXISTS
            ]
        )
        parser.add_argument(
            f'--{constants.GCS_MONGO_BATCH_SIZE}',
            dest=constants.GCS_MONGO_BATCH_SIZE,
            required=False,
            default=constants.MONGO_DEFAULT_BATCH_SIZE,
            help='GCS MONGO Output Batch Size'
        )

        known_args: argparse.Namespace
        known_args, _ = parser.parse_known_args(args)

        return vars(known_args)

    def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:

        logger: Logger = self.get_logger(spark=spark)

        # Arguments
        input_location: str = args[constants.GCS_MONGO_INPUT_LOCATION]
        input_format: str = args[constants.GCS_MONGO_INPUT_FORMAT]
        output_uri:str = args[constants.GCS_MONGO_OUTPUT_URI]
        output_database:str = args[constants.GCS_MONGO_OUTPUT_DATABASE]
        output_collection:str = args[constants.GCS_MONGO_OUTPUT_COLLECTION]
        output_mode:str = args[constants.GCS_MONGO_OUTPUT_MODE]
        batch_size:int = args[constants.GCS_MONGO_BATCH_SIZE]

        ignore_keys = {constants.GCS_MONGO_OUTPUT_URI}
        filtered_args = {key:val for key,val in args.items() if key not in ignore_keys}
        logger.info(
            "Starting GCS to MONGO spark job with parameters:\n"
            f"{pprint.pformat(filtered_args)}"
        )

        # Read
        input_data = ingest_dataframe_from_cloud_storage(spark, args, input_location, input_format, "gcs.mongo.input.")

        # Write
        input_data.write.format(constants.FORMAT_MONGO)\
            .option(constants.MONGO_URL, output_uri) \
            .option(constants.MONGO_DATABASE, output_database) \
            .option(constants.MONGO_COLLECTION, output_collection) \
            .option(constants.MONGO_BATCH_SIZE, batch_size) \
            .mode(output_mode) \
            .save()

File: java/src/main/java/com/google/cloud/dataproc/templates/gcs/GCStoMongo.java
/*
 * Copyright (C) 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataproc.templates.gcs;

import static com.google.cloud.dataproc.templates.util.TemplateConstants.*;

import com.google.cloud.dataproc.templates.BaseTemplate;
import org.apache.commons.lang3.StringUtils;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class GCStoMongo implements BaseTemplate {
  public static final Logger LOGGER = LoggerFactory.getLogger(GCStoMongo.class);

  private final String inputFileLocation;
  private final String inputFileFormat;
  private final String sparkLogLevel;
  private final String mongoUrl;
  private final String mongoDatabase;
  private final String mongoCollection;
  private String mongoBatchSize;
  private String mongoSaveMode;

  public GCStoMongo() {

    inputFileLocation = getProperties().getProperty(GCS_MONGO_INPUT_LOCATION);
    inputFileFormat = getProperties().getProperty(GCS_MONGO_INPUT_FORMAT);
    mongoUrl = getProperties().getProperty(GCS_MONGO_URL);
    mongoDatabase = getProperties().getProperty(GCS_MONGO_DATABASE);
    mongoCollection = getProperties().getProperty(GCS_MONGO_COLLECTION);
    mongoBatchSize = getProperties().getProperty(GCS_MONGO_BATCH_SIZE);
    mongoSaveMode = getProperties().getProperty(GCS_MONGO_OUTPUT_MODE);
    sparkLogLevel = getProperties().getProperty(SPARK_LOG_LEVEL);
  }

  @Override
  public void runTemplate() {

    SparkSession spark = null;

    spark = SparkSession.builder().appName("GCS to Mongo load").getOrCreate();

    // Set log level
    spark.sparkContext().setLogLevel(sparkLogLevel);

    Dataset<Row> inputData = null;
    switch (inputFileFormat) {
      case SPARK_FILE_FORMAT_CSV:
        inputData =
            spark
                .read()
                .format(inputFileFormat)
                .option(SPARK_CSV_HEADER, true)
                .option(SPARK_INFER_SCHEMA, true)
                .load(inputFileLocation);
        break;
      case SPARK_FILE_FORMAT_PARQUET:
        inputData = spark.read().parquet(inputFileLocation);
        break;
      case SPARK_FILE_FORMAT_AVRO:
        inputData = spark.read().format(inputFileFormat).load(inputFileLocation);
        break;
      case SPARK_FILE_FORMAT_JSON:
        inputData = spark.read().json(inputFileLocation);
        break;
      default:
        throw new IllegalArgumentException(
            "Currently avro, parquet, csv and json are the only supported formats");
    }
    inputData
        .write()
        .format(MONGO_FORMAT)
        .option(MONGO_URL, mongoUrl)
        .option(MONGO_DATABASE, mongoDatabase)
        .option(MONGO_COLLECTION, mongoCollection)
        .option(MONGO_BATCH_SIZE, mongoBatchSize)
        .mode(mongoSaveMode)
        .save();
  }

  public void validateInput() {
    if (StringUtils.isAllBlank(inputFileLocation)
        || StringUtils.isAllBlank(inputFileFormat)
        || StringUtils.isAllBlank(mongoUrl)
        || StringUtils.isAllBlank(mongoDatabase)
        || StringUtils.isAllBlank(mongoCollection)) {
      LOGGER.error(
          "{},{},{},{},{} are required parameters. ",
          GCS_MONGO_INPUT_LOCATION,
          GCS_MONGO_INPUT_FORMAT,
          GCS_MONGO_URL,
          GCS_MONGO_DATABASE,
          GCS_MONGO_COLLECTION);
      throw new IllegalArgumentException(
          "Required parameters for GCStoMongo not passed. "
              + "Set mandatory parameter for GCStoMongo template "
              + "in resources/conf/template.properties file.");
    }

    LOGGER.info(
        "Starting GCS to Mongo spark job with following parameters:"
            + "1. {}:{} | "
            + "2. {}:{} | "
            + "3. {}:{} | "
            + "4. {}:{} | "
            + "5. {}:{} | ",
        GCS_MONGO_INPUT_LOCATION,
        inputFileLocation,
        GCS_MONGO_INPUT_FORMAT,
        inputFileFormat,
        GCS_MONGO_URL,
        mongoUrl,
        GCS_MONGO_DATABASE,
        mongoDatabase,
        GCS_MONGO_COLLECTION,
        mongoCollection);
  }
}
